tests/test_pickle_gem.py

```diff
@@ -0,0 +1,24 @@
+from __future__ import absolute_import, print_function, division
+from six.moves import cPickle as pickle, range
+import gem
+import numpy
+import pytest
+
+
+@pytest.mark.parametrize('protocol', range(3))
+def test_pickle_gem(protocol):
+    f = gem.VariableIndex(gem.Indexed(gem.Variable('facet', (2,)), (1,)))
+    q = gem.Index()
+    r = gem.Index()
+    _1 = gem.Indexed(gem.Literal(numpy.random.rand(3, 6, 8)), (f, q, r))
+    _2 = gem.Indexed(gem.view(gem.Variable('w', (None, None)), slice(8), slice(1)), (r, 0))
+    expr = gem.ComponentTensor(gem.IndexSum(gem.Product(_1, _2), (r,)), (q,))
+
+    unpickled = pickle.loads(pickle.dumps(expr, protocol))
+    assert repr(expr) == repr(unpickled)
+
```

Comment:
They won't be equal and I would argue they shouldn't, and this is because of the semantics of running indices. Two running indices are the same only by object identity, so one cannot accidentally create an Index that is equal to an already existing Index... that is only possible by copying a reference. This object identity however infeasible to maintain through the pickling/unpickling operation.

One might also argue that `==` is overly discriminating, and that is true, but changing that would require much more complicated comparison algorithms.